File: Week_12_Day_4.js
/* https://leetcode.com/problems/two-sum/

Given an array of integers, return indices of the
two numbers such that they add up to a specific target.

You may assume that each input would have EXACTLY ONE SOLUTION,
and you may not use the same element twice.

the array is unsorted, contains no floats, and there may be duplicate values

Given arr = [2, 11, 7, 15], target = 9,
Because arr[0] + arr[2] = 2 + 7 = 9
return [0, 2].

example 1
given: [2, 11, 7, 15]
target: 9
output: [0,2]

example 2
given: [3,2,4]
target: 6
output: [1,2]

example 3
given: [3,3]
target: 6
output: [0,1]
*/

// O(n^2) solution
// function twoSums(arr, target) {
//     const resultArr = [];
//     for (let i = 0; i < arr.length; i++) {
//         for (let j = i + 1; j < arr.length; j++) {
//             if (arr[i] + arr[j] === target) {
//                 resultArr.push(i, j);
//             }
//         }
//     }
//     return resultArr;
// }

// O(n)+ solution
// function twoSums(arr, target) {
//     let resultArr = [];
//     let hashmap = {};

//     for (let i = 0; i < arr.length; i++) {
//         hashmap[arr[i]] = i;
//     }

//     for (let j = 0; j < arr.length; j++){
//         let difference = target - arr[j];

//         if(hashmap[difference] !== null && hashmap[difference] !== j){
//             resultArr.push(j, hashmap[difference]);
//             return resultArr;
//         }
//     }

//     return resultArr;
// }

// // O(n) solution (work in progress)
function twoSums(arr, target) {
    let resultArr = [];
    let hashmap = {};

    for (let i = 0; i < arr.length; i++) {
        if(!(hashmap.hasOwnProperty(arr[i]))){
            hashmap[arr[i]] = i;
        }

        let difference = target - arr[i]
        if(hashmap.hasOwnProperty(difference) && hashmap[difference] !== i){
            resultArr.push(hashmap[difference], i);
            return resultArr;
        }
    }

    return resultArr;
}

console.log(twoSums([2, 11, 7, 15], 9)); // [0,2]
console.log(twoSums([3, 2, 4], 6)); // [1,2]
console.log(twoSums([3, 3], 6)); // [0,1]

File: Week_12_Day_4.py
# TwoSum Python

def twoSumSlow(nums, target):
    for i in range(len(nums)):
        j = i + 1 

        while j < len(nums):
            if nums[i] + nums[j] == target:
                return [i, j]
            else:
                j += 1

# print(twoSumSlow([2, 11, 7, 15], 9)) # [0,2]
# print(twoSumSlow([3, 2, 4], 6)) # [1,2]
# print(twoSumSlow([3, 3], 6)) # [0,1]


def twoSumFast(nums, target):
    hashmap = {}

    for i in range(len(nums)):
        if nums[i] not in hashmap:
            hashmap[nums[i]] = i
        
        difference = target - nums[i]
        if difference in hashmap and hashmap[difference] != i:
            return [hashmap[difference], i]

print(twoSumFast([2, 11, 7, 15], 9)) # [0,2]
print(twoSumFast([3, 2, 4], 6)) # [1,2]
print(twoSumFast([3, 3], 6)) # [0,1]